src/core/server/saved_objects/service/lib/collect_multi_namespace_references.ts

```diff
@@ -0,0 +1,306 @@
+  }
+
+  const objectsWithContext = Array.from(
+    inboundReferencesMap.entries()
+  ).map<SavedObjectReferenceWithContext>(([referenceKey, referenceVal]) => {
+    const inboundReferences = Array.from(referenceVal.entries()).map(([objectKey, name]) => {
+      const { type, id } = parseKey(objectKey);
+      return { type, id, name };
+    });
+    const { type, id } = parseKey(referenceKey);
+    const object = objectMap.get(referenceKey);
+    const spaces = object?.namespaces ?? [];
+    return { type, id, spaces, inboundReferences, ...(object === null && { isMissing: true }) };
+  });
+
+  const aliasesMap = await checkLegacyUrlAliases(
+    client,
+    serializer,
+    getIndexForType,
+    objectsWithContext,
+    searchPerPage
+  );
+  const results = objectsWithContext.map((obj) => {
+    const key = `${obj.type}:${obj.id}`;
+    const val = aliasesMap.get(key);
+    const spacesWithMatchingAliases = val && Array.from(val);
+    return { ...obj, spacesWithMatchingAliases };
+  });
+
+  return {
+    objects: results,
+  };
+}
+
+const PIT_KEEP_ALIVE = '1m';
+
+async function checkLegacyUrlAliases(
+  client: RepositoryEsClient,
+  serializer: SavedObjectsSerializer,
+  getIndexForType: (type: string) => string,
+  objects: SavedObjectReferenceWithContext[],
+  perPage = 1000
+) {
+  const filteredObjects = objects.filter(({ spaces }) => spaces.length !== 0);
+  if (!filteredObjects.length) {
+    return new Map<string, Set<string>>();
+  }
+  let pitId: string | undefined;
+  const index = getIndexForType(LEGACY_URL_ALIAS_TYPE);
+  let returnValue: Map<string, Set<string>> | null = null;
+  try {
+    const { body } = await client.openPointInTime({ index, keep_alive: PIT_KEEP_ALIVE });
+    pitId = body.id;
+
+    const query = createAliasQuery(filteredObjects);
+    let lastResultsCount = 0;
+    let from = 0;
+    const aliasesMap = new Map<string, Set<string>>();
+    do {
+      const results = await client.search<SavedObjectsRawDocSource>({
+        // don't specify index, since we are using the pit
+        body: {
+          pit: { id: pitId, keep_alive: PIT_KEEP_ALIVE },
+          query,
+          from,
+        },
+        size: perPage,
+      });
+      const hits = results.body.hits.hits;
+
+      // @ts-expect-error Hit._source may be undefined
+      const aliases = hits.map((hit) => serializer.rawToSavedObject<LegacyUrlAlias>(hit));
+      for (const alias of aliases) {
+        const { sourceId, targetType, targetNamespace } = alias.attributes;
+        const key = `${targetType}:${sourceId}`;
+        const val = aliasesMap.get(key) ?? new Set<string>();
+        val.add(targetNamespace);
+        aliasesMap.set(key, val);
+      }
+
+      lastResultsCount = hits.length;
+      from += lastResultsCount;
+    } while (lastResultsCount >= perPage); // end the loop when there are fewer hits than our perPage size
+    returnValue = aliasesMap;
+  } catch (e) {
+    // do nothing
```

Comment:
I was also on the fence about this too. My rationale here (which I could have described better in this brief comment!) is that searching for potential namespace conflicts is not _essential_ to the `collectMultiNamespaceReferences` function, it just provides us some nice metadata for potential alias conflicts that we can act on before sharing. Nothing catastrophic happens if we managed to not be able to finish the search in `checkLegacyUrlAliases`.

The way I plan to implement handling potential alias conflicts is to introduce client-side constraints to sharing (since the client first collects objects+references and then shares them later on). If the client is aware of potential alias conflicts, then it can prompt the user to do something about it before sharing.

Anyway, if we'd rather code more defensively here I can certainly change this to re-throw the error.

Reply:
I'm leaning towards treating this as an error, but I can see how this could be viewed as an optional step. Silently failing introduces the risk that we won't be able to properly inform the end user about the consequences of their actions